test/python/circuit/test_initializer.py

```diff
@@ -314,6 +315,29 @@ def test_equivalence(self):
 
         self.assertEqual(qc1, qc2)
 
+    def test_max_number_cnots(self):
+        """
+        Check if the number of cnots <= 2^(n+1) - 2n (arXiv:quant-ph/0406176)
+        """
+        num_qubits = 3
+        _optimization_level = 0
+
+        vector = -1 * np.random.rand(2 ** num_qubits)
+        vector[0] = 0
+        vector = vector / np.linalg.norm(vector)
```

Comment:
Depending on the random numbers generated this test may not be very precise. E.g. for some random numbers also the old implementation may have less than `2^(n+1) - 2n` cnots right? Could you maybe fix some numbers which failed for the old implementation and work in this one? 

Other than that this looks good, I think!

Reply:
Hi @Cryoris, thanks for the revision. I included a test with numbers which failed for the old implementation and work in this one.